### PR TITLE
Example of eventual consistency with a distributed tx

### PR DESCRIPTION
**Pull request by MarcinHoppe**
@SzymonPobiega @tmasternak Please review.

/cc @Particular/docs-maintainers

### CODE REVIEW

**Comment by MarcinHoppe** [comment]
Resolves #2482.

File: nservicebus/transports/transactions.md [hunk 1]
```diff
@@ -51,6 +51,22 @@ In this mode handlers will execute inside the `TransactionScope` created by the
...
+1. MSMQ is a local resource and commits very fast. `StartOrder` message is removed from the input queue and `VerifyPayment` is immediately sent to `PaymentService`.
+1. At this point RavenDB transaction received order to commit but has not yet persisted the `OrderSagaData` instance.
+1. `PaymentService` receives `VerifyPayment` message and immediately responds with `CompleteOrder` message to the originating `OrderSaga`.
+1. `OrderSaga` receives the `CompleteOrder` message and attempts to complete the saga.
+1. At this point RavenDB has still not persisted the `OrderSagaData` instance and the `OrderSaga` fails to find an instance of the saga to complete.
+
+In the example above atomicity is preserved from the `OrderSaga` perspective: receiving `StartOrder` message, storing `OrderSagaData` in RavenDB and sending `VerifyPayment` message are one atomic operation. State of the system is not consistent for an outside observer such as the `PaymentService` and results in unexpected behavior: `CompleteOrder` message is sent even though the `OrderSagaData` does not yet exist in RavenDB and cannot be completed.
```
Comment (SzymonPobiega) [inline nservicebus/transports/transactions.md]: Not sure if I could agree that from the perspective or order saga atomicity is preserved. I'd omit this sentence.

File: nservicebus/transports/transactions.md [hunk 2]
```diff
@@ -51,6 +51,22 @@ In this mode handlers will execute inside the `TransactionScope` created by the
...
+
```
Comment (SzymonPobiega) [inline nservicebus/transports/transactions.md]: I would add something about solution: locks. When the storage maintains the locks by itself (SQL Server) nothing needs to be done. When the storage does not (RavenDB, Oracle), marking a message as starting saga will enforce the lock conflict on the SagaID.

File: nservicebus/transports/transactions.md [hunk 3]
```diff
@@ -36,7 +36,7 @@ partial: matrix
...
 
-If required the transaction is escalated to a distributed one (following two-phase commit protocol). That depends on the transport and persistence choice, for example it is not needed when using [SQL Server transport](/nservicebus/sqlserver/) with [NHibernate persistence](/nservicebus/nhibernate/) both using the same database. In this case the ADO.NET driver guarantees that everything happens inside a single database transaction, ACID guarantees are held for the whole processing.
+If required the transaction is escalated to a distributed one (following two-phase commit protocol) depends on the transport and persistence choice. For example it is not needed when using [SQL Server transport](/nservicebus/sqlserver/) with [NHibernate persistence](/nservicebus/nhibernate/) both using the same database. In this case the ADO.NET driver guarantees that everything happens inside a single database transaction, ACID guarantees are held for the whole processing.
```
Comment (ramonsmits) [inline nservicebus/transports/transactions.md]: The sentence doesn't read correctly, maybe:

> If required, the transaction is escalated to a distributed transaction (based on the two-phase commit protocol with MSDTC) if both the transport and persistence support distributed transactions.

Comment (ramonsmits) [inline nservicebus/transports/transactions.md]: >both using the same database

Must be, if both are using the exact same connection string.

File: nservicebus/transports/transactions.md (hunk 1 above)
Comment (ramonsmits) [inline nservicebus/transports/transactions.md]: As V6 is buffered by default none of the message are actually pushed to the transport until all handlers/sagas are completed.

File: nservicebus/transports/transactions.md (hunk 3 above)
Comment (dvdstelt) [inline nservicebus/transports/transactions.md]: Theoretically it doesn't have to be MSDTC. But I've never seen anything else on Windows! :)

File: nservicebus/transports/transactions.md (hunk 1 above)
Comment (dvdstelt) [inline nservicebus/transports/transactions.md]: You say that the state of the system is not consistent for the `PaymentService`, I don't get that. I don't think that service could care less about the state. It just receives a message and then replies?

Comment (dvdstelt) [inline nservicebus/transports/transactions.md]: I think this part is a bit complicated and hard to follow. But after thinking for it a while, I can't really come up with a nice story to simplify it. Perhaps ask someone from story editors or so?

**Comment by weralabaj** [comment]
@MarcinHoppe can I help? Once @tmasternak approves his review, ping me for a quick review.

**Comment by MarcinHoppe** [comment]
I took some time to rethink how to approach this and it occured to me that what I was trying to explain boils down to eventual consistency and not lack of atomicity.

@tmasternak @SzymonPobiega @ramonsmits @dvdstelt @weralabaj Can you take a look at this attempt at an explanation?

File: nservicebus/transports/transactions.md [hunk 4]
```diff
@@ -45,11 +45,26 @@ NOTE: MSMQ will escalate to a distributed transaction right away since it doesn'
...
+**A distributed transaction between the queueing system and the persistent storage guarantees _atomic commits_ but does not guarantee _consistency_.**
+
+Consider a message exchange scenario with a saga that models a simple order lifecycle:
+
+1. `OrderSaga` receives a `StartOrder` message
+1. New `OrderSagaData` instance is created and stored in RavenDB.
+1. `OrderSaga` sends `VerifyPayment` message to `PaymentService`.
+1. NServiceBus completes the distributed transaction and the DTC instructs MSMQ and RavenDB resource managers to commit their local transactions.
+1. `StartOrder` message is removed from the input queue and `VerifyPayment` is immediately sent to `PaymentService`.
+1. RavenDB transaction is committed and changes are written to the transaction log.
```
Comment (tmasternak) [inline nservicebus/transports/transactions.md]: If that would be the case (transaction was committed) than the `OrderSagaData` would be visible for queries made against RavenDB after commit returned. Secondly the transaction log is appended with (probably) Undo/Redo record before commiting - during prepare phase.

File: nservicebus/transports/transactions.md [hunk 5]
```diff
@@ -45,11 +45,26 @@ NOTE: MSMQ will escalate to a distributed transaction right away since it doesn'
...
 
 
-#### Consistency guarantees
+#### Atomicity and consistency guarantees
 
-In this mode handlers will execute inside the `TransactionScope` created by the transport. This means that all the data updates are executed as a whole or rolled back as a whole.
+In this mode handlers will execute inside a `TransactionScope` created by the transport. This means that all the data updates and queue operations are all committed or all rolled back.
 
-Distributed transactions do not guarantee atomicity for an outside observer. For example outgoing messages might be dispatched and processed by some other endpoint before changes to the database get committed. However it is guaranteed that eventually all resources will sync up to reflect the outcome of the transaction.
+**A distributed transaction between the queueing system and the persistent storage guarantees _atomic commits_ but does not guarantee _consistency_.**
```
Comment (tmasternak) [inline nservicebus/transports/transactions.md]: I would still suggest to change _consistency_ to _atomic visibility_. There are many consistency models and it's not obvious which one we have in mind.

File: nservicebus/transports/transactions.md [hunk 6]
```diff
@@ -45,11 +45,26 @@ NOTE: MSMQ will escalate to a distributed transaction right away since it doesn'
...
+1. `PaymentService` receives `VerifyPayment` message and immediately responds with a `CompleteOrder` message to the originating `OrderSaga`.
+1. `OrderSaga` receives the `CompleteOrder` message and attempts to complete the saga.
+1. `OrderSaga` queries RavenDB to find the `OrderSagaData` instance to complete.
+1. RavenDB returns an empty result set and `OrderSaga` fails to complete. This may be caused either by the DTC or the snapshot isolation guarantee provided by the underlying storage engine (Esent).
+
+In the example above the `TransactionScope` guarantees atomicity for the `OrderSaga`: consuming the incoming `StartOrder` message, storing `OrderSagaData` in RavenDB and sending the outgoing `VerifyPayment` message are _committed_ as one atomic operation. _Consistency_, however, is not immediately preserved because the saga data is not yet available for reading even though the incoming message has already been processed. `OrderSaga` is thus _eventually consistent_: the `CompleteOrder` message needs to be [retried](/nservicebus/recoverability/) until RavenDB successfuly returns an `OrderSagaData` instance.
```
Comment (tmasternak) [inline nservicebus/transports/transactions.md]: One more argument to rethink usage of _consistency_. AFAIK _consistency_ as a concept is used to describe possible histories/behaviors of a storage system under more than one, possibly concurrent transactions.  In this sample all three resources i.e. input_queue+RavenDB+output_queue are _the_ storage and the transaction is made of operations that are executed when incoming message is processed. So it would make sense to talk about _consistency_ in context of multiple DTC transactions and not in context of operations inside single DTC transaction.

File: nservicebus/transports/transactions.md [hunk 7]
```diff
@@ -36,7 +36,7 @@ partial: matrix
...
 
-If required the transaction is escalated to a distributed one (following two-phase commit protocol). That depends on the transport and persistence choice, for example it is not needed when using [SQL Server transport](/nservicebus/sqlserver/) with [NHibernate persistence](/nservicebus/nhibernate/) both using the same database. In this case the ADO.NET driver guarantees that everything happens inside a single database transaction, ACID guarantees are held for the whole processing.
+If required, the transaction is escalated to a distributed transaction (following two-phase commit protocol coordinated by MS DTC) if both the transport and the persistence support it. A fully distributed transaction is not always required, for example when using [SQL Server transport](/nservicebus/sqlserver/) with [NHibernate persistence](/nservicebus/nhibernate/) both using the same database connection string. In this case the ADO.NET driver guarantees that everything happens inside a single database transaction and ACID guarantees are held for the whole processing.
```
Comment (weralabaj) [inline nservicebus/transports/transactions.md]: That's assuming we use the same connection string for transport & persistence, not sure if that's mentioned earlier ;)

File: nservicebus/transports/transactions.md [hunk 8]
```diff
@@ -45,11 +45,26 @@ NOTE: MSMQ will escalate to a distributed transaction right away since it doesn'
...
-#### Consistency guarantees
+#### Atomicity and consistency guarantees
 
-In this mode handlers will execute inside the `TransactionScope` created by the transport. This means that all the data updates are executed as a whole or rolled back as a whole.
+In this mode handlers will execute inside a `TransactionScope` created by the transport. This means that all the data updates and queue operations are all committed or all rolled back.
 
-Distributed transactions do not guarantee atomicity for an outside observer. For example outgoing messages might be dispatched and processed by some other endpoint before changes to the database get committed. However it is guaranteed that eventually all resources will sync up to reflect the outcome of the transaction.
+**A distributed transaction between the queueing system and the persistent storage guarantees _atomic commits_ but does not guarantee _consistency_.**
+
+Consider a message exchange scenario with a saga that models a simple order lifecycle:
```
Comment (weralabaj) [inline nservicebus/transports/transactions.md]: Consider a message exchange scenario with a saga that models a simple order lifecycle for MSMQ transport and RavenDB persistence:

File: nservicebus/transports/transactions.md [hunk 9]
```diff
@@ -45,11 +45,26 @@ NOTE: MSMQ will escalate to a distributed transaction right away since it doesn'
...
+1. New `OrderSagaData` instance is created and stored in RavenDB.
+1. `OrderSaga` sends `VerifyPayment` message to `PaymentService`.
+1. NServiceBus completes the distributed transaction and the DTC instructs MSMQ and RavenDB resource managers to commit their local transactions.
+1. `StartOrder` message is removed from the input queue and `VerifyPayment` is immediately sent to `PaymentService`.
+1. RavenDB transaction is committed and changes are written to the transaction log.
+1. `PaymentService` receives `VerifyPayment` message and immediately responds with a `CompleteOrder` message to the originating `OrderSaga`.
+1. `OrderSaga` receives the `CompleteOrder` message and attempts to complete the saga.
+1. `OrderSaga` queries RavenDB to find the `OrderSagaData` instance to complete.
+1. RavenDB returns an empty result set and `OrderSaga` fails to complete. This may be caused either by the DTC or the snapshot isolation guarantee provided by the underlying storage engine (Esent).
+
```
Comment (weralabaj) [inline nservicebus/transports/transactions.md]: Given that there are problems with RavenDB DTC implementation https://github.com/Particular/PlatformDevelopment/issues/1275 could we use SQL as an example? Or that's specifically required?

File: nservicebus/transports/transactions.md (hunk 7 above)
Comment (MarcinHoppe) [inline nservicebus/transports/transactions.md]: I think this is what this sentence says:

> for example when using A with B, **both** using the same database connection string.

File: nservicebus/transports/transactions.md (hunk 9 above)
Comment (MarcinHoppe) [inline nservicebus/transports/transactions.md]: We've seen it with RavenDB. I'll add a note that it is also possible with lower isolation levels on relational databases.

**Comment by MarcinHoppe** [comment]
@tmasternak @weralabaj Another batch of improvements.

**Comment by weralabaj** [comment]
@tmasternak if you're ok with changes, go ahead and merge

**Comment by MarcinHoppe** [comment]
@weralabaj I addressed final (hopefully!) remarks from @tmasternak.
